### PR TITLE
Deal with Datagram msg leak false-positive

### PR DESCRIPTION
**Pull request by rakhimov**
**By submitting this pull-request, I confirm the following:**

- I have read and understood the [Contributing Guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.

*Place an X inside the bracket (or click the box in preview) to confirm*
- [x] I confirm.

---
This is a Coverity reported leak,
which seems to be false positive.
The current refactoring is to make the code easier on Coverity.

---
This change is [Reviewable](https://reviewable.io/reviews/monero-project/kovri/326)

### CODE REVIEW

File: src/client/datagram.cc [hunk 1]
```diff
@@ -59,43 +59,43 @@ void DatagramDestination::SendDatagramTo(
...
+  I2NPMessage* msg = CreateDataMessage(buf, len + headerLen, fromPort, toPort);
+  std::shared_ptr<const i2p::data::LeaseSet> remote
+      = m_Owner.FindLeaseSet(ident);
+  if (remote) {
     m_Owner.GetService().post(
-        std::bind(
-          &DatagramDestination::SendMsg,
-          this, msg, remote));
-  else
+        [this, msg, remote] { SendMsg(msg, remote); });
```
Comment (rakhimov) [inline src/client/datagram.cc]: Is there any possibility that this callback is never executed?
(if it is not executed, then the message will not get deleted.)

File: src/client/datagram.cc [hunk 2]
```diff
@@ -59,43 +59,43 @@ void DatagramDestination::SendDatagramTo(
...
   } else {
     m_Owner.Sign(buf1, len, signature);
   }
-  auto msg =
-    CreateDataMessage(buf, len + headerLen, fromPort, toPort);
-  auto remote = m_Owner.FindLeaseSet(ident);
-  if (remote)
+  I2NPMessage* msg = CreateDataMessage(buf, len + headerLen, fromPort, toPort);
+  std::shared_ptr<const i2p::data::LeaseSet> remote
+      = m_Owner.FindLeaseSet(ident);
```
Comment (anonimal) [inline src/client/datagram.cc]: I don't see a need to refactor-out ```auto``` (to do so is counterproductive for abstraction and should be refactored back).

Comment (rakhimov) [inline src/client/datagram.cc]: Please see #325

File: src/client/datagram.cc (hunk 1 above)
Comment (anonimal) [inline src/client/datagram.cc]: No, not that I can see. If lease-set is available, great, if not, it's requested below.

**Comment by anonimal** [comment]
Thanks @rakhimov. I'll push a coverity build after this is merged.

**Comment by anonimal** [comment]
Referencing #263.

**Pull request merged**